### PR TITLE
[Auditbeat] fim(kprobes): enrich file events by coupling add_process_metadata processor

### PR DESCRIPTION
**Pull request by pkoutsovasilis**
## Proposed commit message

This PR adds reporting of  process.group.id, process.group.name and process.entity_id in `add_process_metadata` processor. Also it changes the factory of MetricSets to allow the latter to specify Processors after successful instantiation; this is required as FIM has 3 different available backends, namely `fsnotify`, `kprobes`, `ebpf` and only the `kprobes` one requires to have `add_process_metadata` processor. Utilising the former, `kprobes` backend always adds a properly configured `add_process_metadata` processor. As a result, enriching kprobes file events with process-related data exhibits the same robustness levels of the current `add_process_metadata` processor. However, the current design is aligned with @nick-alayil and the `sec-linux-platform` which plan to increase the robustness of `add_process_metadata` processor in a separate effort.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Depends on https://github.com/elastic/beats/pull/38328

## Related issues

- Relates https://github.com/elastic/integrations/issues/7401

## Screenshots

image

### CODE REVIEW

**Review by belimawr — APPROVED** [review]

**Review by andrewkroh — COMMENTED** [review]
That's a pretty clean approach you took to add processors to MetricSets. I think we just need a few more details in the interface godocs as mentioned in my comment.

------------

I tend to view Beats as being the building block to getting data ingested how you want it. So I am not in favor of turning extra fields on by default. For example, I would rather see events starting as empty documents and the inputs contributing the raw data. Anything else is opt-in by the user through processors whether that be agent metadata, host metadata, process metadata, etc (but I'm probably not the typical user).

Today's situation is that we have a lot of things that are on by default (or not even possible to disable like `host.name` or `agent.*` or even docker metadata when run under Elastic Agent) so if you want to opt-out your only solution is to delete the fields with a processor (which is wasteful).

But I understand the desire to have everything just work out of the box. It's hard to serve the two personas that want it to just work and want to be a composable ingest tool. So allow me to mention an alternative.

1. Document within Auditbeat's FIM guide how to add the enrichment processor yourself. This gives the user full control over what fields are they wish to add. e.g.

```yaml
processors:
  - if.and:
      - equals.event.module: file_integrity
      - not.has_fields: [process.name]
      - has_fields: [process.pid]
    then:
      add_process_metadata:
        match_pids:
        - process.pid
```

2. Add the processor to the Fleet FIM integration configuration. For a Fleet integration, the batteries should be included.

File: auditbeat/module/file_integrity/metricset.go [hunk 1]
```diff
@@ -105,6 +114,13 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 		log:           logger,
 	}
 
+	// reader supports  a processor
+	if rWithProcessor, ok := r.(eventProducerWithProcessor); ok {
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply auditbeat/module/file_integrity/metricset.go:118]: unfortunately the Run method of the processor has a singature of that `Run(event *beat.Event)` where the eventreader of FIM module is a `mb.Event` and I haven't find an easy way to transition back and forth between them

File: metricbeat/mb/module/factory.go [hunk 2]
```diff
@@ -31,6 +31,11 @@ type Factory struct {
 	options  []Option
 }
 
+// metricSetWithProcessors is an interface to check if a MetricSet has directly attached Processors
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply metricbeat/mb/module/factory.go:34]: nice catch @andrewkroh ty ty! I run some more experiment, and indeed you are 100% right, the processor is getting closed at each reload, I will adjust my code to instantiate a processor each time the kprobe reader is getting instantiated

File: auditbeat/module/file_integrity/eventreader_kprobes.go [hunk 3]
```diff
@@ -39,6 +44,41 @@ type kProbesReader struct {
 	log     *logp.Logger
 
 	parsers []FileParser
+
+	processor beat.Processor
+}
+
+var processMetadataProcessorOnce = sync.OnceValues(func() (beat.Processor, error) {
+	config, err := conf.NewConfigFrom(mapstr.M{
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply auditbeat/module/file_integrity/eventreader_kprobes.go]: I did introduce a New* function with config option, happy to hear your thoughts on it @andrewkroh 🙂

File: metricbeat/mb/module/factory.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply metricbeat/mb/module/factory.go:35]: I tested this with multiple sequential changes (disable -> enable -> disable -> enable -> change monitor path) with dynamic config reloading enabled, and in every case the processor gets closed and then (with the current code) re-instantiated alongside the kprobes reader

**Review by andrewkroh — APPROVED** [review]
LGTM

**Pull request merged**